### ISSUE TITLE
[ban-types] `{}` is banned on the left of `extends` in ternary statements where it *has* a meaning

### ISSUE THREAD

**Issue by phryneas** [typescript-eslint/typescript-eslint#2370]
**Repro**

```TS
declare type EmptyPropertyKeys<T extends {} /* here the rule would make sense */> = {
    [K in keyof T]: {} /* here it definitely works as intended, ban-types should not warn */ extends T[K] ? K : never;
}[keyof T];

const someObject = {foo: {asd: 1}, bar: {}, baz: 1, bingo: null};
type X = EmptyPropertyKeys<typeof someObject> // results in only the string "bar", as expected
```

**Expected Result**

No warning in the `{} extends ... ? ... : ...` position, as it does what it should do in that position as seen in the reproduction above.
(here is a [playground](https://www.typescriptlang.org/play?#code/FAegVMAqAWCWDOACAJgUwLYHsB28AuATgIZ6pJ7QmIWqIBmmANo5gO6zYDmiZjHeiAgFdGteNDZIhAB0Qc52AMaYCBVIoHTGRRWQBcwYABEcAcgFD4tAAYBvAL7XERJEWoBPaagB0iO4+cNISJmd0R0VCJcRAAiKLDsHABabBE+cUQANxChVBjvYCTEAEk6RHdMIURWKIE3PE9aCKiObjjsMMwAIwArdTwYgBpyysRpAm6iLsYwmuwBawAldRVkAB58AlbhoWwAa0TWbAA+Jw58SOQCotKRqrm6jy9wyOxW2PisnLzhiqrxybTWa1Py7A5sbBnXCkIhXRAACl4-AAAg0vPBFFtpHgkkj5iAulEkmiyABKYBgECGNCKbRqJ60ACi6Gx7gAChMvAQGgBpVDueBrSA8AAepGwyCQDkQ4EQ0FQ9JoghEtFYlUYyHCRD2YlQuFolOOiAAvIhbMBEJbEABtHkKRA6iplSAAXT0ZvsMrAcoVtFgAjQdA4-tQM2qKj2riQ-D1aE1lNF4sliEgtpdiAA-Ig7e7sKhMgqANzAezWx2YZ0u4vAZTQxDwTARADyvX6JrNDEw7tsLmQ7oAjPZhoSCN2h4hCQAvAfD1pdxCpZj2YskxAADXbzNZHMwXN5-MFJIr9cbqBbfQ0RpAIEEZBEeGj2AnRAIwxcoq8GlQyCAA))

**Actual Result**

> Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead. (eslint@typescript-eslint/ban-types)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.5.0` |
| `@typescript-eslint/parser`        | `3.5.0` |
| `TypeScript`                       | `3.9.6` |
| `ESLint`                           | `7.4.0` |
| `node`                             | `12.18.2` |
| `npm`                              | `6.14.5` |
| `yarn`                              | `2.1.1` |

**Comment by bradzacher** [comment]
The `ban-types` rule is intentionally simple. It is a generic rule for outright banning a type across your codebase.

If you are intentionally using a type that the rule bans, you must use an eslint disable comment (or alternately reconfigure the rule).

**Comment by phryneas** [comment]
@bradzacher might I suggest to "weaken" the statement error message for this case in some way then?

Right now, it contains statements about the type system that do not hold true in every situation - and I've seen a lot of devs just taking those "error popup" informations for gospel, not consulting the TypeScript docs or actually playing around with it to determine if the statement even holds true. (Not just in RL, but also a lot on Reactiflux).

Honestly, I have to admit that I don't know how to word it better, though :/

**Comment by bradzacher** [comment]
We write rules here for the majority case. The default options for all of our rules cover those cases.

`{}` is a tricky one. In 99.9999% of cases, the error is correct.
In just two cases I know that it's not: `extends {}` where it actually means empty object, and `& {}` where it is just a noop.

Both of these cases are rare, and generally indicate some advanced types.
In a case like this (being an advanced case), I'd expect the user to be switch on enough to see the error, question it and test, and disable as appropriate.

In all other cases I'm okay with the user blindly following the rule, because it will prevent weak typesafety from leaking into the codebase.
